p2p/p2p.go

```diff
@@ -116,18 +116,58 @@ func adaptDiscRouting(udpNode *discover.UDPv5, peers map[peer.ID]enode.Node) pee
 			return peer.AddrInfo{}, errors.New("peer not resolved")
 		}
 
-		mAddr, err := multiAddrFromIPPort(resolved.IP(), resolved.TCP())
-		if err != nil {
-			return peer.AddrInfo{}, err
+		var mAddrs []ma.Multiaddr
+
+		// If tcp port is 0, this node isn't bound to a port.
+		if resolved.TCP() != 0 {
+			mAddr, err := multiAddrFromIPPort(resolved.IP(), resolved.TCP())
+			if err != nil {
+				return peer.AddrInfo{}, err
+			}
+			mAddrs = append(mAddrs, mAddr)
+		}
+
+		// Add any circuit relays
+		for _, relay := range udpNode.Relays {
+			if relay.Enode.TCP() == 0 {
+				continue
+			}
+
+			relayAddr, err := multiAddrViaRelay(relay, peerID)
+			if err != nil {
+				return peer.AddrInfo{}, err
+			}
+			mAddrs = append(mAddrs, relayAddr)
+		}
```

Comment:
here we basically tell libp2p that it should do circuit relay if the above resolved address isn't reachable.